### ISSUE TITLE
Problem with tmap_animation and a suggestion

### ISSUE THREAD

**Issue by dhersz** [mtennekes/tmap#446]
I have come across a problem with `tmap_animation` when a `restart_delay` is set and the number of frames is big. The issue might be related to the fact that my operating system is Windows, but here is a reproducible code (I have manually edited the reprex to make it more readable - the dots are replacing messages similar to the ones that are shown):

``` r
n <- length(unique(spData::world$name_long))
tmap_options(max.categories = n, limits = c(facets.view = 4, facets.plot = n))

animated_world_map <- tm_shape(spData::world) +
  tm_polygons(col = "name_long") + tm_facets(by = "name_long", ncol = 1, nrow = 1)

tmap_animation(animated_world_map, filename = str_c("animated_world_map.gif"), width = 1500, height = 1500, delay = 10, restart.delay = 25)
#> Warning in syscall("convert -version"): execução de 'convert -version' falhou
#> com código de erro 4
#> Warning in syscall(paste(program, " -loop ", loop, " ", part1,
#> " ", part2, : execução de 'magick convert -loop 0 -delay 10 C:
#> \Users\Usuario\AppData\Local\Temp\RtmpUn6RZJ//tmap_plots/plot001.png -delay 10
#> C:\Users\Usuario\AppData\Local\Temp\RtmpUn6RZJ//tmap_plots/plot002.png -delay 10
#> .
#> .
#> .
#> C:\Users\Usuario\AppData\Local\Temp\RtmpUn6RZJ//tmap_plots/plot099.png -delay 10
#> C:\Users\Usuario\AppData\Local\Temp\RtmpUn6RZJ//tmap_plots/plot100.png -delay 10
#> C:\Users\Usuario\AppData\Local\Temp\RtmpUn6R
#> Animation saved to C:\Users\Usuario\AppData\Local\Temp\RtmpozcE32\reprex1ee412d14f4b\animated_world_map.gif
```

Even though the warning message shows that the animation has been saved, that's not true. No gif is created after running this code.

The issue is related to the length of the string sent to the system, which exceeds the maximum characteres it can handle. Below I reproduce the same error with the same approach used to create the animation in `tmap_animation` (I'm using some frames I created with another code, but I have made them available [in this repository](https://github.com/dhersz/tmap_animation_example) if you wanna try it out):

``` r
syscall <- if (.Platform$OS.type == "unix") system else shell

checkIM <- syscall("convert -version")
#> Warning in syscall("convert -version"): execução de 'convert -version' falhou
#> com código de erro 4
if (checkIM==0) {
  program <- "convert"
} else {
  checkIM2 <- syscall("magick convert -version")
  if (checkIM2!=0) {
    stop("Could not find ImageMagick. Make sure it is installed and included in the systems PATH")  
  } else program <- "magick convert"
}

d <- str_c(getwd(), "/frames_long")

files <- list.files(d)
k <- length(files)

filename = str_c(getwd(), "/animation.gif")
delay <- 10
restart.delay <- 25

part1 <- str_c(str_c("-delay ", delay, " ", d, "/", files[1L:(k-1)]), collapse = " ")
part2 <- str_c("-delay ", restart.delay, " ", d, "/", files[k])
call_msg <- str_c(program, " -loop 0 ", part1, " ", part2, " \"", filename, "\"")

syscall(call_msg, mustWork = TRUE)
#> Error in syscall(call_msg, mustWork = TRUE): execução de 'magick convert -loop 0 -delay 10 C:/Users/Usuario/Documents/frames_long/frame_0001.png -delay 10 C:/Users/Usuario/Documents/frames_long/frame_0002.png ................ -delay 10 C:/Users/Usuario/Documents/frames_long/frame_0126.png -delay 10 C:/Users/Usuario/Documents/frames_long/frame_0127.png -delay 10 C:/Users/Usuario
```

Making the following changes to how `part1` and `part2` are created solves this error (I changed `delay` and `restart.delay` to make it more visible that it works). 
p.s: I still get the error code 4 warning -reprex doesn't show it-, but the animation is created anyway.

``` r
delay <- 2
restart.delay <- 500

part1 <- str_c("-delay ", delay, " ", d, "/*.png")
part2 <- str_c("-delay ", restart.delay - delay, " ", d, "/", files[k])
call_msg <- str_c(program, " -loop 0 ", part1, " ", part2, " \"", filename, "\"")

syscall(call_msg, mustWork = TRUE)
```

![animation](https://user-images.githubusercontent.com/1557047/81596129-eafdc380-9399-11ea-8dff-4612f615cc18.gif)

I wasn't sure If I should have made a pull request, so I decided to report the issue first. Let me know if you see any problems with the approach I showed.

**Edit:** 

I also have a suggestion for `tmap_animation`. I'm not sure if this is related to my setup or not, but gif making seems to take very long after a certain number of frames. I show some timings below (`frames_short` is a folder with 100 frames, `frames_long` with 200 and `frames_longer` with 250):

```r
create_gif <- function(filename, frames_dir, delay, restart.delay) {
    
    syscall <- if (.Platform$OS.type == "unix") system else shell
    
    checkIM <- syscall("convert -version")
    if (checkIM==0) {
        program <- "convert"
    } else {
        checkIM2 <- syscall("magick convert -version")
        if (checkIM2!=0) {
            stop("Could not find ImageMagick. Make sure it is installed and included in the systems PATH")
        } else program <- "magick convert"
    }
    
    d <- str_c(getwd(), "/", frames_dir)
    
    files <- list.files(d)
    k <- length(files)
    
    filename = str_c(getwd(), "/", filename)
    
    part1 <- str_c("-delay ", delay, " ", d, "/*.png")
    part2 <- str_c("-delay ", restart.delay - delay, " ", d, "/", files[k])
    call_msg <- str_c(program, " -loop 0 ", part1, " ", part2, " \"", filename, "\"")
    
    syscall(call_msg, mustWork = TRUE)
    
}

system.time(create_gif("animation.gif", "frames_short", 2, 50))
#>      user system elapsed
#>      0.02   0.03   14.97

system.time(create_gif("animation.gif", "frames_long", 2, 50))
#>      user system elapsed
#>      0.03   0.03   94.49

system.time(create_gif("animation.gif", "frames_longer", 2, 50))
#>      user system  elapsed
#>      0.03   0.03   704.07
```

The frames I'm using belong to a project with 2000+ total frames. Initially I planned to animate them with a gif, but after some long hours with no result I tried to make a movie (mp4), which turned out to be much faster (I used `ffmpeg`, which requires its installation and inclusion to system PATH, at least with Windows):

```r
create_movie <- function(filename, frames_dir, delay, restart.delay, width = 2000, height = 1300) {
  
  syscall <- if (.Platform$OS.type == "unix") system else shell
  
  d <- str_c("C:/Users/Usuario/Documents", "/", frames_dir)
  
  files <- list.files(d)
  k <- length(files)
  
  filename = str_c("C:/Users/Usuario/Documents", "/", filename)
  if (file.exists(filename)) file.remove(filename)
  
  duration <- (k * delay + restart.delay - delay) / 100
  fps <- 100 / delay
  
  call_msg <- str_c(
    "ffmpeg",
    " -f lavfi -i nullsrc=s=", width, "x", height, ":d=", duration, ":r=", fps,
    " -framerate ", fps,
    " -i ", d, "/frame_%04d.png",
    " -filter_complex \"[0:v][1:v]overlay[video]\" -map \"[video]\" -shortest",
    " -pix_fmt yuv420p ", filename
  )
  
  syscall(call_msg, mustWork = TRUE)
  
}

system.time(create_movie("animation.mp4", "frames_short", 2, 50))
#>      user system elapsed
#>      0.00   0.05    4.89

system.time(create_movie("animation.mp4", "frames_long", 2, 50))
#>      user system elapsed
#>      0.01   0.03   11.34

system.time(create_movie("animation.mp4", "frames_longer", 2, 50))
#>      user system elapsed
#>      0.02   0.00   13.76
```

The files sizes are also much smaller (250 frames mp4 164 KB vs 11.8 MB gif).

So my suggestion is to add an animation method that returns a movie instead of a gif. I think it could be useful for some larger animations such as mine.

Also, is it possible to add a progress bar to `tmap_animation`?
When I was still trying to create the gif with it I wasn't sure if my waiting was due to the map plotting or the gif making (turns out it was due to both), so I think that adding some way of measuring progress such as messages like `Plotting map 1 out of n... Plotting map 2 out of n... Plotting map n out of n... Creating animated gif...` could be useful (I don't know how viable it is though).

Sorry for the long post, got a bit too excited I guess. Cheers

**Comment by mtennekes** [comment]
Thanks a lot for this great contribution! Could you make a pull request?

It's good to have both options: animation and movie. I suggest we add a new function `tmap_movie`. I think it would make sense to add an argument "frame.rate", since this is more common than "delay" for movies. If we choose to have the same defaults we can specify 
`tmap_movie(... , delay = 40, frame.rate = 100 / delay)`

**Comment by dhersz** [comment]
@mtennekes I'm happy that you found the suggestions useful. I'm quite busy these last few days, but I'll try to make a pull request on the weekend! Cheers

**Comment by mtennekes** [comment]
@dhersz I plan to submit a new version of tmap (3.1) to CRAN next week. If you are able to make a pull request soon, I will include it in this version. No worries if you don't make it; then it will be included in the next version.

**Comment by dhersz** [comment]
@mtennekes Ok, cool. Unfortunately I probably I won't be able to make it in the next two weeks or so, but I plan to do it afterwards. Sorry for this "delay", but I don't really have much experience with making pull requests and coding for libraries, so I prefer doing it whenever I have more time to spend on it to prevent any errors that could be caused by the rush.

**Comment by mtennekes** [comment]
Thanks a lot. That is a good approach!

**Comment by Nowosad** [comment]
@dhersz I would suggest using the **av** package (https://cran.r-project.org/web/packages/av/index.html) instead of directly calling FFmpeg. I used it recently in one of my projects and it worked like a charm.

**Comment by mtennekes** [comment]
Yes, `av` works like charm indeed!

I've improved `tmap_animation`, but we can further improve it.

* Do we still want to use ImageMagick directly (as in the current version) or shall we switch to av/ffmpeg completely? I'm open to the latter, although I found out that the quality of the gif created with av is not as good:

![Dutch_provinces](https://user-images.githubusercontent.com/2444081/92485381-2941ad00-f1eb-11ea-9d3c-e7ccb857df5a.gif)

Is this caused by how ffmpeg is configured (see also my next point)?
* I have used ffmpeg before, but I'm not an expert. Which settings/filters are best for tmap and can we set them via the av packge? For instance, the lines from @dhersz's suggested code above `" -filter_complex \"[0:v][1:v]overlay[video]\" -map \"[video]\" -shortest",
    " -pix_fmt yuv420p ", filename`
* Shall we have two functions `tmap_animation` that is aimed for short gifs and `tmap_movie` that is aimed for movies (eg mp4) with a large number of frames? Other suggestions are welcome.

**Comment by tim-salabim** [comment]
Maybe consider [gifski](https://gif.ski/) through https://github.com/r-rust/gifski ?

**Comment by Nowosad** [comment]
Awesome @mtennekes. 
My thoughts: 
1. I would just keep the `tmap_animation()` function for both gifs and videos.
2. I second @tim-salabim suggestion about gifski - it usually returns gifs with visibly smaller sizes than ImageMagick without decreasing the quality.

**Comment by mtennekes** [comment]
Done! Thanks for the wonderful packages `av` and `gifski` @jeroen!

Other improvements:
* It also supports a list of tmap objects
* I changed the default dpi value to 100. It was 300, which is also the default of `tmap_save`. However, animations are typically saved at a much lower resolution than static maps, so 100 would make more sense to me. Besides the tmap option `output.dpi` (still used by `tmap_save`, I've added the option `output.dpi.animation`.
* The examples have been updated. For instance:

![Life expectancy](https://user-images.githubusercontent.com/2444081/92619446-ca9b3280-f2c1-11ea-837d-0860ac5e5513.gif)
